File: src/main/java/ch/ethz/covspectrum/config/JwtRequestFilter.java
package ch.ethz.covspectrum.config;

import ch.ethz.covspectrum.service.JwtUserDetailsService;
import ch.ethz.covspectrum.util.JwtTokenService;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;


@Component
public class JwtRequestFilter extends OncePerRequestFilter {

    private final JwtUserDetailsService jwtUserDetailsService;

    private final JwtTokenService jwtTokenService;


    public JwtRequestFilter(JwtUserDetailsService jwtUserDetailsService, JwtTokenService jwtTokenService) {
        this.jwtUserDetailsService = jwtUserDetailsService;
        this.jwtTokenService = jwtTokenService;
    }


    /**
     * This function looks for the "Authorization" header and validates the JWT token if one is present.
     *
     * TODO Better handling of error cases such as expired tokens. See: JwtTokenValidationService#validateJwtToken
     */
    @Override
    protected void doFilterInternal(
            HttpServletRequest request,
            HttpServletResponse response,
            FilterChain chain
    ) throws ServletException, IOException {
        String requestTokenHeader = request.getHeader("Authorization");
        if (requestTokenHeader != null && requestTokenHeader.startsWith("Bearer ")) {
            String jwtToken = requestTokenHeader.substring(7); // Removes "Bearer " prefix
            String username = jwtTokenService.getUsername(jwtToken);

            UserDetails userDetails = this.jwtUserDetailsService.loadUserByUsername(username);
            if (jwtTokenService.validateToken(jwtToken, userDetails)) {
                var usernamePasswordAuthenticationToken = new UsernamePasswordAuthenticationToken(
                        userDetails, null, userDetails.getAuthorities());
                usernamePasswordAuthenticationToken
                        .setDetails(new WebAuthenticationDetailsSource().buildDetails(request));
                SecurityContextHolder.getContext().setAuthentication(usernamePasswordAuthenticationToken);
            }
        }
        chain.doFilter(request, response);
    }

}

File: src/main/kotlin/ch/ethz/covspectrum/config/JwtRequestFilter.kt
package ch.ethz.covspectrum.config

import ch.ethz.covspectrum.service.JwtTokenService
import ch.ethz.covspectrum.service.JwtUserDetailsService
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse


@Component
class JwtRequestFilter(
    private val jwtUserDetailsService: JwtUserDetailsService,
    private val jwtTokenService: JwtTokenService
): OncePerRequestFilter() {

    /**
     * This function looks for the "Authorization" header and validates the JWT token if one is present.
     *
     * TODO Better handling of error cases such as expired tokens. See: JwtTokenService#validateJwtToken
     */
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        // First check if a JWT token is provided as query param
        var jwtToken = request.getParameter("jwt")

        // If not, check the Authorization header
        if (jwtToken == null || jwtToken.isBlank()) {
            val requestTokenHeader = request.getHeader("Authorization")
            if (requestTokenHeader != null && requestTokenHeader.startsWith("Bearer ")) {
                jwtToken = requestTokenHeader.substring(7) // Removes "Bearer " prefix
            }
        }
        if (jwtToken != null) {
            try {
                val restrictionEndpoint = jwtTokenService.getRestrictionEndpoint(jwtToken)
                if (restrictionEndpoint != null && restrictionEndpoint != request.requestURI) {
                    response.status = 401
                    return
                }
                val username = jwtTokenService.getUsername(jwtToken)
                val userDetails = jwtUserDetailsService.loadUserByUsername(username)
                if (jwtTokenService.validateToken(jwtToken, userDetails)) {
                    val usernamePasswordAuthenticationToken = UsernamePasswordAuthenticationToken(
                        userDetails, null, userDetails.authorities
                    )
                    usernamePasswordAuthenticationToken.details = WebAuthenticationDetailsSource().buildDetails(request)
                    SecurityContextHolder.getContext().authentication = usernamePasswordAuthenticationToken
                }
            } catch (e: Exception) {
                response.status = 401
                return
            }
        }
        filterChain.doFilter(request, response)
    }

}
